cmd/shared/main.go

```diff
@@ -0,0 +1,77 @@
+package main
+
+/*
+struct StoredSecret{
+    char *Name;
+    char *ID;
+    char *Provider;
+};
+*/
+import "C"
+import (
+	"fmt"
+	"github.com/cyberark/secretless-broker/internal/app/secretless"
+	"github.com/cyberark/secretless-broker/internal/pkg/plugin"
+	"github.com/cyberark/secretless-broker/pkg/secretless/config"
+	"crypto/sha1"
+	plugin_v1 "github.com/cyberark/secretless-broker/pkg/secretless/plugin/v1"
+)
+
+func NewStoredSecret(ref C.struct_StoredSecret) config.StoredSecret {
+	return config.StoredSecret{
+		Name:     C.GoString(ref.Name),
+		ID:       C.GoString(ref.ID),
+		Provider: C.GoString(ref.Provider),
+	}
+}
+
+func GetSecrets(secrets []config.StoredSecret) (map[string][]byte, error)  {
+	// Load all internal Providers
+	providerFactories := make(map[string]func(plugin_v1.ProviderOptions) (plugin_v1.Provider, error))
+	for providerID, providerFactory := range secretless.InternalProviders {
+		providerFactories[providerID] = providerFactory
+	}
+
+	resolver := plugin.NewResolver(providerFactories, nil, nil)
+
+	return resolver.Resolve(secrets)
+}
+
+//export GetSecret
+func GetSecret(cRef C.struct_StoredSecret) (*C.char) {
+	ref := NewStoredSecret(cRef)
+	secrets, err := GetSecrets([]config.StoredSecret{ref})
+	if err != nil {
+		fmt.Println("Error fetching secret")
+		return C.CString("")
+	}
+	return C.CString(string(secrets[ref.Name]))
+}
+
+//export NativePassword
+func NativePassword(cRef C.struct_StoredSecret, salt *C.char) (*C.char) {
```

Comment:
Is this MySQL-specific?

Reply:
can we just use this method here? https://github.com/cyberark/secretless-broker/blob/master/internal/app/secretless/handlers/mysql/protocol/protocol.go#L713

Reply:
same functionality but different signatures. i'm trying to avoid the Go strings because I can't control when their memory is freed.

Reply:
so, if we change the input for the existing one to `[]byte` instead of string, you would use that? bc I am 100% for that, rather than having the functionality defined in 2 places

Reply:
the method converts it to a byte array very early on anyway

Reply:
fair point. done